### DISCUSSION TITLE
[Windows DirectML] Traceback that doesn't appear in sdnext.log

### DISCUSSION THREAD

**Discussion by CrimsonCuttle** [vladmandic/sdnext#2238] [Q&A] [answered]
Attempting to run `webui.bat --use-directml` ends with a traceback error which doesn't appear on sdnext.log. Attached are the log file and the traceback after which I copied from the command prompt.
The error in specific seems to be the `AttributeError: partially initialized module 'modules.shared' has no attribute 'log' (most likely due to a circular import)`.

[sdnext.log](https://github.com/vladmandic/automatic/files/12697468/sdnext.log)
[aaagh.txt](https://github.com/vladmandic/automatic/files/12697470/aaagh.txt)

**Comment by brknsoul** [comment]
Sounds like your initial start of webui.bat was without `--use-directml`

try;
`webui.bat --use-directml --reinstall`

failing that, you could try deleting venv, and re-running the above.
failing that, you could just nuke the whole lot and start again (copy the models and output folders out, so you don't lose your models or genned images.)

**Reply by CrimsonCuttle** [reply]
This unfortunately did not work. I'd also like to point out that in the initial log, it does mention directML so I think it did register the command.

**Comment by vladmandic** [comment, marked as answer]
directml was installed, but it failed during initialization. unfortunately, it failed in such a place that shared logger was not initialized yet and thats why the strange traceback. i'll rearrange the code so logger is available in time to capture the error. unfortunately, not much more i can do right now than try to capture the error.

new code is now in dev branch if you can try it out.

**Reply by CrimsonCuttle** [reply]
So after wiping and installing the dev branch, it seems to have actually ran mostly fine, without the traceback. Though, torch-directml doesn't seem to install, so it just seems to default to CPU.
[sdnext.log](https://github.com/vladmandic/automatic/files/12710493/sdnext.log)

**Reply by vladmandic** [reply]
log says clearly "no matching distribution found" - there is no `torch-directml` for python 3.11, you need to use python 3.10 for directml.

**Reply by CrimsonCuttle** [reply]
Oop. Thought I was using 3.10. Any way to specify which python version to use? I've got so many.

**Reply by vladmandic** [reply]
set `PYTHON` env variable to path to your python executable.

**Reply by CrimsonCuttle** [reply]
Ah, thank you. Using 3.10.6 it seems to run just fine until it says "Device:" and then once again throws an exception which does not appear in the log:
[sdnext.log](https://github.com/vladmandic/automatic/files/12710701/sdnext.log)
[graaarg.txt](https://github.com/vladmandic/automatic/files/12710702/graaarg.txt)

**Reply by vladmandic** [reply]
ah, magic of dev branch - i was in the middle of refactoring upscalers. will fix in few min, this has nothing to do with directml.

**Reply by CrimsonCuttle** [reply]
I see. I will check back in a few then, thank you.

**Reply by CrimsonCuttle** [reply]
Just downloaded latest ver and it's working great! Thank you!
